IDiamondDragon/nodejs-aws-shop-be
# Task 9 (Backend For Frontend)

### **THE BFF SERVICE WAS TERMINATED AS A RESULT THE SITE DOESN'T WORK** 

## TASK 9.2
---
**My score: 7 points;**

**Frontend:** 
PR: https://github.com/IDiamondDragon/nodejs-aws-fe/pull/7
Frontend: http://d3jfdstsfa8xga.cloudfront.net (because of we can get data from the cart api only through the http requests you need to open the site http://d3jfdstsfa8xga.cloudfront.net through http if you can't do it thorough Chrome you can do it through Mozila for example)

## Evaluation criteria (each mark includes previous mark criteria)
---

Provide your reviewers with the following information:
- link to the repo
- **product-service** service API endpoint URL
Product service url: https://ppauvpo36j.execute-api.eu-west-1.amazonaws.com/dev/products
- example of the **create product** API call with all needed information: URL, payload, headers, etc.
To send request by using Postman when you will set body you need select raw body and type raw body must be JSON(the same for all post requests).

![image](https://user-images.githubusercontent.com/17451770/102011126-bc92fd00-3d5b-11eb-8dad-e6b34beb6a10.png)

Example product:
{
    "title": "Viny Pooh",
    "description": "Viny Pooh",
    "price": "4",
    "count": "4"
}

- **CART** service API endpoint URL
Cart service url: http://idiamonddragon-cart-api-dev.eu-west-1.elasticbeanstalk.com/api/profile/cart


- **bff-service** service URL
Product service url: http://idiamonddragon-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
Cart service url: http://idiamonddragon-bff-api-develop.eu-west-1.elasticbeanstalk.com/cart
- example how to call **product-service** and **CART** services via **bff-service** service URL   
get all products
![image](https://user-images.githubusercontent.com/17451770/102011339-e3056800-3d5c-11eb-856d-9b9d2505f062.png)

get one products by id

![image](https://user-images.githubusercontent.com/17451770/102011366-fd3f4600-3d5c-11eb-991e-7dab048992ef.png)

get cart

![image](https://user-images.githubusercontent.com/17451770/102011387-1fd15f00-3d5d-11eb-9eef-c324aa189a31.png)

 put cart

![image](https://user-images.githubusercontent.com/17451770/102011644-f6193780-3d5e-11eb-9725-409ef132bcaf.png)

Example cart:

{
    "items": [
        {
            "count": 1,
            "product": {
                "id": "989368f3-5353-4a41-8a6a-64bf46a303e1", 
                "title": "yyyy",
                "description": "yyyy", 
                "price": 4, 
                "count": 4
            }
        }
    ]
}

delete cart

![image](https://user-images.githubusercontent.com/17451770/102011663-23fe7c00-3d5f-11eb-80d6-f76188acc43b.png)

checkout cart

when there aren't products in cart

![image](https://user-images.githubusercontent.com/17451770/102011739-adae4980-3d5f-11eb-88c2-f05aedf5bb9d.png)



when there are:

![image](https://user-images.githubusercontent.com/17451770/102011748-c0c11980-3d5f-11eb-9dee-413ec2606569.png)

---
* **3** - A working and correct **express** application should be in the **bff-service** folder. Reviewer can start this application locally with any valid configuration in the **.env** file and this application should works as described in the task 9.1 **- Done**
* **5** - The **bff-service** should be deployed with Elastic Beanstalk. The **bff-service** call should be redirected to the appropriate service : **product-service** or **CART**. The response from the **bff-service** should be the same as if **product-service** or **CART** services were called directly. **- Done**



## Additional (optional) tasks
---
* **+1** - Add a cache at the **bff-service** level for a request to the **getProductsList** function of the **product-service**. The cache should expire in 2 minutes.  **- Done**
How to test:
  * Get products list
  * Create new product
  * Get products list - result shouldn’t have new product
  * Wait more than 2 minutes
  * Get products list - result should have new product
* **+1** - Use **NestJS** to create **bff-service** instead of **express**   **- Done**
Folder: bff-nestjs-service

**My score: 7 points;**
